### ISSUE TITLE
Missing files when deploying using build-context local 

### ISSUE THREAD

**Issue by juan-s-cruz** [TheRacetrack/racetrack#555]
When using "--build-context local" to deploy a job, all files with file formats stated in the .gitignore are not included in the image which is normally desired. However, this also ignores files that were added with "add -f" to the repo and may be required by the job.

**Comment by LookACastle** [comment]
The reason it works like that is that the local build context is implemented for rapid developments sake, I.E, debugging code and such without having to care about the quality as you explore, because it's not going to end up in the git history. When all is said and done though, Racetrack does expect your code to live in a git repository eventually - so obeying `.gitignore` is very much on purpose.

I can't really find a purpose for `git add -f`, so I'd be happy to learn what the use case here is - it seems antithetical to good developer ergonomics to have a non-representative `.gitignore` file.

**Comment by juan-s-cruz** [comment]
It is exactly testing where I am at. 

My situation is, say you have several files (e.g. pickles) that generally not all go to the repo until they are tested and decisions have been made on which one should be commited. The file format might generally be excluded, because many such objects exist in the development process. One of those was chosen to be included in the repo. 
  
Now I have new changes and I do not want to commit to the repo yet, but I would like to test the deployment, so I opted for the local build-context option. Then the deployment tries to look for the necessary file but of course it was not included.

Do you have any suggestions on what would be better practice?

**Comment by LookACastle** [comment]
That setup sounds like you should have feature branches or similar instead, that way other people can review and test the code as well. Then your `.gitignore` could be representative of the real code (since it'd just be the `.gitignore` of the feature branch, you could change it however you wanted), and this problem would be bypassed.

Besides, `git add -f` still adds the files to the repository as far as I can read, so one of us is misunderstanding what `git add -f` does.

I'll get a hold of the lead dev, but from my understanding this is Racetrack working as intended, and your setup not aligning with that. When it comes to this project though, I've mostly just lurked the issues without touching the code for a while now though, so I may very well be wrong in that assessment.

**Comment by iszulcdeepsense** [comment]
@juan-s-cruz I understand your problem. However, as @LookACastle said, local build context is implemented for rapid development sake. It's an experimental feature and it should be used with caution. Keep in mind that it packs the whole repository state, so if you're going to put large pickles there, the things may get heavy and sluggish pretty fast. Thus, the recommended way is to use **git build context** instead.
Then, you won't face this issue, because files seen by git will be always the same no matter where it's accessed, including files added forcefully. I presume since you did `git add -f`, you have probably also made a commit, so it's all in the git already and using git build context would fit to your workflow.
Although it makes local build context a bit inconsistent, at this point we don't plann to scan the whole git repository in search of the forcefully added files, this level of integration with git would be too intrusive in my opinion.
In a nutshell, if you want to be in sync with git, use "git" build context.

**Comment by iszulcdeepsense** [comment]
git has a pattern for exceptions in .gitignore: https://stackoverflow.com/a/2415885/6772197
I think we could support that, that wouldn't be much hassle and it doesn't require relying on git being installed.
